rclcpp/src/rclcpp/executor.cpp

```diff
@@ -679,24 +676,13 @@ Executor::get_next_ready_executable(AnyExecutable & any_executable)
   }
 
   if (!valid_executable) {
-    size_t current_timer_index = 0;
-    while (true) {
-      auto [timer, timer_index] = wait_result_->peek_next_ready_timer(current_timer_index);
-      if (nullptr == timer) {
-        break;
-      }
-      current_timer_index = timer_index;
+    while (auto timer = wait_result_->next_ready_timer()) {
       auto entity_iter = current_collection_.timers.find(timer->get_timer_handle().get());
       if (entity_iter != current_collection_.timers.end()) {
         auto callback_group = entity_iter->second.callback_group.lock();
-        if (callback_group && !callback_group->can_be_taken_from()) {
+        if (!callback_group || !callback_group->can_be_taken_from()) {
           continue;
         }
-        // At this point the timer is either ready for execution or was perhaps
-        // it was canceled, based on the result of call(), but either way it
-        // should not be checked again from peek_next_ready_timer(), so clear
-        // it from the wait result.
-        wait_result_->clear_timer_with_index(current_timer_index);
```

Comment:
We made the change to peek and conditionally clear after a lot of iterations, and I don't fully understand why this needs to be changed back to not peek before clearing. Can you give a better explanation as to why this isn't needed any more? Is it related to the logic change around checking the callback groups?

Specifically, if this function is called multiple times before the `wait_result_` is discarded due to a new call to wait on the wait set, we need to ensure timers are not passed over due to callback group constraints on one call only to be ignore on subsequent calls. If we could ensure that wait is called again (and the `wait_result_` is discarded) every time a callback group became available to be taken from, then this wouldn't be needed, but I think for performance the idea is that wait results are fully processed before waiting again, in which case iterating over a timer, not calling it because of its callback group, but marking it as nullptr in the wait result anyways, would result in it getting skipped when it shouldn't.

Reply:
I did not readd this change for now, so we can go ahead with this PR. 

Anyway, currently the behavior of the WaitResult is inconsistent. 
The timers are the only objects, that feature the 'do not clear' logic if the callback group
is in used. I actually see not a real reason for this, perhaps to make some test happy ?

All other objects in the WaitResult clear on iteration. Therefore I would suggest, that we
either also clear the timers, or move the callback_group->can_be_taken_from() check inside
of the WaitResult and do not clear if blocked, to get a consistent behavior.

Reply:
> That's why timers have a different API here.

I'm confused about this statement. Timers stay ready as well on subsequent waits.